src/Context/ContextApi.php

```diff
@@ -76,26 +115,111 @@ private function getXSLContextSourceHeader() {
     public function uploadContext(UploadContextParameters $params)
     {
         $requestData = $this->getDefaultRequestData('multipart', $params->exportToArray());
-        $requestData['headers']['X-SL-Context-Source'] = $this->getXSLContextSourceHeader();
 
         return $this->sendRequest('contexts', $requestData, self::HTTP_METHOD_POST);
     }
 
-  /**
-   * Match context.
-   *
-   * @param $contextUid
-   * @return array
-   * @throws \Smartling\Exceptions\SmartlingApiException
-   */
+    /**
+     * Match context.
+     *
+     * @param $contextUid
+     * @return array
+     * @throws \Smartling\Exceptions\SmartlingApiException
+     */
     public function matchContext($contextUid)
     {
         $endpoint = vsprintf('contexts/%s/match/async', $contextUid);
         $requestData = $this->getDefaultRequestData('form_params', []);
         $requestData['headers']['Content-Type'] = 'application/json';
-        $requestData['headers']['X-SL-Context-Source'] = $this->getXSLContextSourceHeader();
 
         return $this->sendRequest($endpoint, $requestData, self::HTTP_METHOD_POST);
     }
 
+    /**
+     * Upload and match async.
+     *
+     * @param \Smartling\Context\Params\UploadContextParameters $params
+     * @return array
+     * @throws \Smartling\Exceptions\SmartlingApiException
+     */
+    public function uploadAndMatchContext(UploadContextParameters $params)
+    {
+        $requestData = $this->getDefaultRequestData('multipart', $params->exportToArray());
+
+        return $this->sendRequest('contexts/upload-and-match-async', $requestData, self::HTTP_METHOD_POST);
+    }
+
+    /**
+     * Get missing resources.
+     *
+     * @param MissingResourcesParameters|null $params
+     * @return array
+     * @throws \Smartling\Exceptions\SmartlingApiException
+     */
+    public function getMissingResources(MissingResourcesParameters $params = null) {
+        $requestData = $this->getDefaultRequestData('query', is_null($params) ? [] : $params->exportToArray());
+
+        return $this->sendRequest('missing-resources', $requestData, self::HTTP_METHOD_GET);
+    }
+
+    /**
+     * Get all missing resources.
+     *
+     * @return array
+     * Contains next keys:
+     *  - "items": array of missing resources
+     *  - "all": boolean which indicates whether function has read all the
```

Comment:
Honestly i a little bit confused by such design. Usually if you can set timeout then you should expect exception. I understand that in our case it's ok to load list of resources partially. So let's keep the implementation as is while we try to build something and get feedback.

Reply:
The main idea was to get at least something from this method. If it would exception the we will not be able to read items which were read. Or we would need to extend SmartlingException and implement to set/get read items which is weird (Exception driven development). So I decided to use such design.

"All or nothing" vs "all other at least something"

Reply:
* implement method

Reply:
"or at least something"

Sorry, I'm from phone